IBC.md

```diff
@@ -0,0 +1,412 @@
+
+### Channel Close
+
+A contract may request to close a channel that belongs to it via the following
+`CosmosMsg`:
+
+```rust
+pub enum IbcMsg {
+    /// This will close an existing channel that is owned by this contract.
+    /// Port is auto-assigned to the contracts' ibc port
+    CloseChannel { channel_id: String },
+}
+```
+
+Once a channel is closed, whether due to an IBC error, at our request, or at the
+request of the other side, the following callback is made on the contract, which
+allows it to take appropriate cleanup action:
+
+```rust
+#[entry_point]
+pub fn ibc_channel_close(
+    deps: DepsMut,
+    env: Env,
+    channel: IbcChannel,
+) -> StdResult<IbcBasicResponse> { }
+```
+
+### Packet Lifecycle
+
+Unfortunately the
+[IBC spec on Pakcet Lifecycle](https://github.com/cosmos/ibc/tree/master/spec/core/ics-004-channel-and-packet-semantics#packet-flow--handling)
+is missing all useful diagrams, but it may provide some theoretical background
+for this text if you wish to look.
+
+In short, IBC allows us to send packets from chain A to chain B and get a
+response from them. The first step is the contract/module in chain A requesting
+to send a packet. This is then relayed to chain B, where it "receives" the
+packet and calculates an "acknowledgement" (which may contain a success result
+or an error message, as opaque bytes to be interpretted by the sending
+contract). The acknowledgement is then relayed back to chain A, completing the
+cycle.
+
+In some cases, the packet may never be delivered, and if it is proven not to be
+delivered before the timeout period, this can abort the packet, calling the
+"timeout" handler on chain A. In this case, chain A sends and later gets
+"timeout". No "receive" nor "acknowledgement" callbacks are ever executed.
+
+#### Sending a Packet
+
+In order to send a packet, a contract can simply return `IbcMsg::SendPacket`
+along with the channel over which to send the packet (which you saved in
+`ibc_channel_connect`), as well as opaque data bytes to be interpreted by the
+other side. You must also return a timeout either as block height or block time
+of the remote chain, just like in the ICS20 `Transfer` messages above:
+
+```rust
+pub enum IbcMsg {
+    /// Sends an IBC packet with given data over the existing channel.
+    /// Data should be encoded in a format defined by the channel version,
+    /// and the module on the other side should know how to parse this.
+    SendPacket {
+        channel_id: String,
+        data: Binary,
+        /// block height after which the packet times out.
+        /// at least one of timeout_block, timeout_timestamp is required
+        timeout_block: Option<IbcTimeoutBlock>,
+        /// block timestamp (nanoseconds since UNIX epoch) after which the packet times out.
+        /// See https://golang.org/pkg/time/#Time.UnixNano
+        /// at least one of timeout_block, timeout_timestamp is required
+        timeout_timestamp: Option<u64>,
```

Comment:
Random idea: We could make this type save (only allow 3 of the 4 states) by creating a dedicated IbcTimeout enum with 3 cases (timestamp, height, both).

Reply:
This is a bit more annoying from the Go side of things, but maybe it does make sense.
When I thought it was time or block, I was going to use an enum. With all 3 I figured that was a pain.

Reply:
```rust
pub enum IbcTimeout {
    Timestamp(u64),
    Block(IbcTimeoutBlock),
    Both{
        timestamp: u64,
        block: IbcTimeoutBlock,
    }
}
```

Would give us eg. `timeout: IbcTimeout::Timestamp(1234)` instead of `timeout_block: None, timeout_timestamp: Some(1234)` and makes sense for `IbcTimeout::Block` as well. `IbcTimeout::Both` would be rather verbose...

```
timeout: IbcTimeout::Both{
  timestamp: 1234,
  block: IbcTimeoutBlock { revision: 1, height: 5000 },
}
```

vs

```
timeout_timestamp: Some(1234),
timeout_block: Some({ revision: 1, height: 5000 }),
```

on reflection, I guess this is not much worse in this case for rust, and cleaner for the others. The Go complexity we just code once (and not every contract), so, yeah I will go with this suggestion.

Reply:
They can all be positional:

```rust
pub enum IbcTimeout {
    Timestamp(u64),
    Block(IbcTimeoutBlock),
    Both(u64, IbcTimeoutBlock),
}
```

Reply:
I made this adjustment in  f13fc2e  and  298a421 happy for feedback on that.

Reply:
> They can all be positional:

Mixed positionals are very difficult to parse in Go (or many other languages) - mixed type arrays are not friendly in most languages. I would avoid that.

Reply:
Ahh, I see. Makes sense, good point.